### PR TITLE
Retrieve `MIN_{CANDIDATE,DELEGATOR}_STAKE` from contract, not from env.

### PR DESCRIPTION
**Pull request by Heimdell**
## Motivation

We should retrieve `MIN_CANDIDATE_STAKE` and `MIN_DELEGATOR_STAKE` from contract, not from env.

## Changelog

- [x] `EpochCounter` is `ChainReader`;
- [x] Made `ChainReader` retrieve both stakes from chain;
- [ ] Replace usage of stakes from env with stakes from chain.

## Checklist for your PR

  - [ ] I added an entry to `CHANGELOG.md` with this PR
  - [ ] If I added new functionality, I added tests covering it.
  - [ ] If I fixed a bug, I added a regression test to prevent the bug from silently reappearing again.
  - [ ] I checked whether I should update the docs and did so if necessary

### CODE REVIEW

File: apps/explorer/lib/explorer/staking/chain_reader.ex [hunk 1]
```diff
@@ -81,19 +86,28 @@ defmodule Explorer.Staking.EpochCounter do
...
     with data <- get_epoch_info(),
          {:ok, [epoch_num]} <- data["stakingEpoch"],
-         {:ok, [epoch_end_block]} <- data["stakingEpochEndBlock"] do
+         {:ok, [epoch_end_block]} <- data["stakingEpochEndBlock"],
+         {:ok, [min_delegator_stake]} <- data["getDelegatorMinStake"],
+         {:ok, [min_candidate_stake]} <- data["getCandidateMinStake"] do
       :ets.insert(@table_name, {@epoch_key, epoch_num})
       :ets.insert(@table_name, {@epoch_end_key, epoch_end_block})
+      :ets.insert(@table_name, {@min_delegator_stake_key, min_delegator_stake})
+      :ets.insert(@table_name, {@min_candidate_stake_key, min_candidate_stake})
```
Comment (goodsoft) [inline apps/explorer/lib/explorer/staking/chain_reader.ex]: It is possible to insert multiple entries using one `:ets.insert` call by passing it a list of tuples.

File: apps/explorer/lib/explorer/staking/chain_reader.ex [hunk 2]
```diff
@@ -12,31 +12,36 @@ defmodule Explorer.Staking.EpochCounter do
   @table_name __MODULE__
   @epoch_key "epoch_num"
   @epoch_end_key "epoch_end_block"
+  @min_candidate_stake_key "min_candidate_stake"
+  @min_delegator_stake_key "min_delegator_stake"
 
-  @doc "Current staking epoch number"
-  def epoch_number do
+  def get_parameter_or_default(param, default) do
```
[the remark this answers is not in the document]
Comment (Heimdell) [reply apps/explorer/lib/explorer/staking/chain_reader.ex]: I think, access to the non-existent ETS table is an error. What about letting it crash right there, if there's no such table?

[the remark this answers is not in the document]
Comment (Heimdell) [reply apps/explorer/lib/explorer/staking/chain_reader.ex]: The table name is `__MODULE__`, so I think I can remove the `if` altogether.

File: apps/explorer/lib/explorer/staking/contract_state.ex [hunk 3]
```diff
@@ -12,31 +12,36 @@ defmodule Explorer.Staking.EpochCounter do
   @table_name __MODULE__
   @epoch_key "epoch_num"
   @epoch_end_key "epoch_end_block"
+  @min_candidate_stake_key "min_candidate_stake"
+  @min_delegator_stake_key "min_delegator_stake"
 
-  @doc "Current staking epoch number"
-  def epoch_number do
+  defp get(param, default) do
```
Comment (goodsoft) [inline apps/explorer/lib/explorer/staking/contract_state.ex]: Nitpick: it's more common to place private methods _under_ the public ones.

File: apps/explorer/lib/explorer/staking/contract_state.ex [hunk 4]
```diff
@@ -12,31 +12,36 @@ defmodule Explorer.Staking.EpochCounter do
   @table_name __MODULE__
   @epoch_key "epoch_num"
   @epoch_end_key "epoch_end_block"
+  @min_candidate_stake_key "min_candidate_stake"
+  @min_delegator_stake_key "min_delegator_stake"
```
Comment (goodsoft) [inline apps/explorer/lib/explorer/staking/contract_state.ex]: I'm not sure why strings are used for ETS keys and why are they in module attributes, but can't we just use atoms directly?
cc @saneery

Comment (saneery) [reply apps/explorer/lib/explorer/staking/contract_state.ex]: I don't mind

File: apps/explorer/lib/explorer/staking/contract_state.ex [hunk 5]
```diff
@@ -10,33 +16,35 @@ defmodule Explorer.Staking.EpochCounter do
   alias Explorer.SmartContract.Reader
 
   @table_name __MODULE__
-  @epoch_key "epoch_num"
-  @epoch_end_key "epoch_end_block"
+  @epoch_key :epoch_num
+  @epoch_end_key :epoch_end_block
+  @min_candidate_stake_key :min_candidate_stake
+  @min_delegator_stake_key :min_delegator_stake
+  @token_contract_address :token_contract_address
```
Comment (goodsoft) [inline apps/explorer/lib/explorer/staking/contract_state.ex]: Just use these atoms directly.

File: apps/explorer/lib/explorer/staking/contract_state.ex [hunk 6]
```diff
@@ -1,6 +1,12 @@
-defmodule Explorer.Staking.EpochCounter do
+defmodule Explorer.Staking.ContractState do
   @moduledoc """
-  Fetches current staking epoch number and the epoch end block number.
+  Fetches:
+  1) current staking epoch number;
+  2) the epoch end block number;
+  3) minimal candidate stake;
+  4) minimal delegate stake;
+  6) token contract address.
```
Comment (goodsoft) [inline apps/explorer/lib/explorer/staking/contract_state.ex]: :thinking:

File: apps/explorer/lib/explorer/staking/contract_state.ex [hunk 7]
```diff
@@ -10,33 +16,35 @@ defmodule Explorer.Staking.EpochCounter do
...
-
-        _ ->
-          0
-      end
-    end
+    get(@epoch_key, 0)
   end
 
-  @doc "Block number on which will start new epoch"
+  @doc "Current staking epoch number"
```
Comment (saneery) [inline apps/explorer/lib/explorer/staking/contract_state.ex]: `@doc` does not match to function

File: apps/explorer/lib/explorer/staking/contract_state.ex [hunk 8]
```diff
@@ -81,19 +89,42 @@ defmodule Explorer.Staking.EpochCounter do
     {:noreply, state}
   end
 
-  defp fetch_epoch_info do
+  defp get(param, default) do
+    if :ets.info(@table_name) != :undefined do
```
Comment (saneery) [inline apps/explorer/lib/explorer/staking/contract_state.ex]: Maybe we should return default if the table doesn't exist?

**Pull request merged**